internal/linker/test/link_node_modules.spec.ts

```diff
@@ -150,55 +310,361 @@ describe('link_node_modules', () => {
-    expect(fs.readFileSync(
-               path.join(process.env['TEST_TMPDIR']!, workspace, 'node_modules', 'a', 'index.js'),
-               'utf-8'))
-        .toEqual('/*a*/exports = {}');
-    expect(fs.readFileSync(
-               path.join(process.env['TEST_TMPDIR']!, workspace, 'node_modules', 'b', 'index.js'),
-               'utf-8'))
-        .toEqual('/*b*/exports = {}');
-    expect(
-        fs.readFileSync(
-            path.join(process.env['TEST_TMPDIR']!, workspace, 'node_modules', 'b', 'b', 'index.js'),
-            'utf-8'))
-        .toEqual('/*b/b*/exports = {}');
-    expect(fs.readFileSync(
-               path.join(
-                   process.env['TEST_TMPDIR']!, workspace, 'node_modules', '@foo', 'c', 'index.js'),
-               'utf-8'))
-        .toEqual('/*@foo/c*/exports = {}');
-    expect(fs.readFileSync(
-               path.join(
-                   process.env['TEST_TMPDIR']!, workspace, 'node_modules', '@foo', 'c', 'c',
-                   'index.js'),
-               'utf-8'))
+    expect(readWorkspaceNodeModules('a', 'index.js')).toEqual('/*a*/exports = {}');
+    expect(readWorkspaceNodeModules('b', 'index.js')).toEqual('/*b*/exports = {}');
+    expect(readWorkspaceNodeModules('b', 'b', 'index.js')).toEqual('/*b/b*/exports = {}');
+    expect(readWorkspaceNodeModules('@foo', 'c', 'index.js')).toEqual('/*@foo/c*/exports = {}');
+    expect(readWorkspaceNodeModules('@foo', 'c', 'c', 'index.js'))
         .toEqual('/*@foo/c/c*/exports = {}');
-    expect(fs.readFileSync(
-               path.join(
-                   process.env['TEST_TMPDIR']!, workspace, 'node_modules', '@foo', 'c', 'c', 'c',
-                   'c', 'index.js'),
-               'utf-8'))
+    expect(readWorkspaceNodeModules('@foo', 'c', 'c', 'c', 'c', 'index.js'))
         .toEqual('/*@foo/c/c/c/c*/exports = {}');
-    expect(fs.readFileSync(
-               path.join(
-                   process.env['TEST_TMPDIR']!, workspace, 'node_modules', '@foo', 'd', 'index.js'),
-               'utf-8'))
-        .toEqual('/*@foo/d*/exports = {}');
-    expect(fs.readFileSync(
-               path.join(
-                   process.env['TEST_TMPDIR']!, workspace, 'node_modules', '@foo', 'd', 'bar',
-                   'index.js'),
-               'utf-8'))
+    expect(readWorkspaceNodeModules('@foo', 'd', 'index.js')).toEqual('/*@foo/d*/exports = {}');
+    expect(readWorkspaceNodeModules('@foo', 'd', 'bar', 'index.js'))
         .toEqual('/*@foo/d/bar*/exports = {}');
-    expect(fs.readFileSync(
-               path.join(
-                   process.env['TEST_TMPDIR']!, workspace, 'node_modules', '@foo', 'd', 'bar',
-                   'fum', 'far', 'index.js'),
-               'utf-8'))
+    expect(readWorkspaceNodeModules('@foo', 'd', 'bar', 'fum', 'far', 'index.js'))
         .toEqual('/*@foo/d/bar/fum/far*/exports = {}');
   });
 
+  it('should handle first-party packages with sibling directories', async () => {
```

Comment:
These tests talking to the real FS are slower. It seems like you add all the same coverage below as in the IN/OUT unit tests above. Do we really need both? Do these new tests below add coverage that we don't get from the ones above?

Reply:
They go one step further and test the use of that OUT tree, concurrently traversing it and applying it to the FS.

They could also be run with both the old and new code so I could assure I'm actually fixing things, the test failed before and passes now etc.

We could potentially remove a few now that we're (I hope) more confident everything is working. Only keep the ones that concentrate more on applying the tree to the FS...